1.9.0-jumbo-1+bleeding-84a4aeb20 - wpapsk-opencl broken after driver/cuda update

OpenCL broken (after driver/cuda update) on some hash modes:

word.list content:
hashcat!

pmkid.john content:
`2582a8281bf9d4308d6f5731d0e61c61*4604ba734d4e*89acf0e761f4*ed487162465a774bfba60eb603a39f3a`

both taken from https://hashcat.net/wiki/doku.php?id=example_hashes
```
$ john -w:word.list --format=wpapsk-opencl pmkid.john
Device 1@tux1: NVIDIA GeForce GTX 1080 Ti
Using default input encoding: UTF-8
Loaded 1 password hash (wpapsk-opencl, WPA/WPA2/PMF/PMKID PSK [PBKDF2-SHA1 OpenCL])
5 errors generated.
Options used: -I opencl -cl-mad-enable -DSM_MAJOR=6 -DSM_MINOR=1 -D__GPU__ -DDEVICE_INFO=524306 -D__SIZEOF_HOST_SIZE_T__=8 -DDEV_VER_MAJOR=520 -DDEV_VER_MINOR=56 -D_OPENCL_COMPILER -DHASH_LOOPS=105 -DITERATIONS=4095 -DPLAINTEXT_LENGTH=63 -DV_WIDTH=1 /usr/share/john/opencl/wpapsk_kernel.cl
Build log: In file included from <kernel>:12:
opencl/opencl_sha2_ctx.h:122:24: error: passing '__generic uchar *' (aka '__generic unsigned char *') to parameter of type 'const uchar *' (aka 'const unsigned char *') changes address space of pointer
                _sha256_process(ctx, ctx->buffer);
                                     ^~~~~~~~~~~
opencl/opencl_sha2_ctx.h:40:51: note: passing argument to parameter 'data' here
void _sha256_process(SHA256_CTX *ctx, const uchar data[64]) {
                                                  ^
opencl/opencl_sha2_ctx.h:130:24: error: passing 'const __generic uchar *' (aka 'const __generic unsigned char *') to parameter of type 'const uchar *' (aka 'const unsigned char *') changes address space of pointer
                _sha256_process(ctx, input);
                                     ^~~~~
opencl/opencl_sha2_ctx.h:40:51: note: passing argument to parameter 'data' here
void _sha256_process(SHA256_CTX *ctx, const uchar data[64]) {
                                                  ^
opencl/opencl_sha2_ctx.h:291:24: error: passing '__generic uchar *' (aka '__generic unsigned char *') to parameter of type 'const uchar *' (aka 'const unsigned char *') changes address space of pointer
                _sha512_process(ctx, ctx->buffer);
                                     ^~~~~~~~~~~
opencl/opencl_sha2_ctx.h:209:51: note: passing argument to parameter 'data' here
void _sha512_process(SHA512_CTX *ctx, const uchar data[128]) {
                                                  ^
opencl/opencl_sha2_ctx.h:299:24: error: passing 'const __generic uchar *' (aka 'const __generic unsigned char *') to parameter of type 'const uchar *' (aka 'const unsigned char *') changes address space of pointer
                _sha512_process(ctx, input);
                                     ^~~~~
opencl/opencl_sha2_ctx.h:209:51: note: passing argument to parameter 'data' here
void _sha512_process(SHA512_CTX *ctx, const uchar data[128]) {
                                                  ^
<kernel>:549:15: error: passing '__generic uchar *' (aka '__generic unsigned char *') to parameter of type 'uchar *' (aka 'unsigned char *') changes address space of pointer
        SHA256_Final(mac, &ctx);
                     ^~~
opencl/opencl_sha2_ctx.h:145:25: note: passing argument to parameter 'output' here
void SHA256_Final(uchar output[32], SHA256_CTX *ctx) {
                        ^

Error building kernel /usr/share/john/opencl/wpapsk_kernel.cl. DEVICE_INFO=524306
0: OpenCL CL_BUILD_PROGRAM_FAILURE (-11) error in opencl_common.c:1296 - clBuildProgram

```

Additional information about john:
$ john --list=build-info
Version: 1.9.0-jumbo-1+bleeding-84a4aeb20 2022-10-17 14:03:56 +0200
Build: linux-gnu 64-bit x86_64 AVX AC MPI + OMP OPENCL
SIMD: AVX, interleaving: MD4:3 MD5:3 SHA1:1 SHA256:1 SHA512:1
System-wide exec: /usr/bin
System-wide home: /usr/share/john
Private home: ~/.john
CPU tests: AVX
CPU fallback binary: john-non-avx
$JOHN is /usr/share/john/
Format interface version: 14
Max. number of reported tunable costs: 4
Rec file version: REC4
Charset file version: CHR3
CHARSET_MIN: 1 (0x01)
CHARSET_MAX: 255 (0xff)
CHARSET_LENGTH: 24
SALT_HASH_SIZE: 1048576
SINGLE_IDX_MAX: 2147483648
SINGLE_BUF_MAX: 4294967295
Effective limit: Number of salts vs. SingleMaxBufferSize
Max. Markov mode level: 400
Max. Markov mode password length: 30
gcc version: 12.2.0
GNU libc version: 2.36 (loaded: 2.36)
OpenCL headers version: 1.2
Crypto library: OpenSSL
OpenSSL library version: 01010111f
OpenSSL 1.1.1q  5 Jul 2022
GMP library version: 6.2.1
File locking: fcntl()
fseek(): fseek
ftell(): ftell
fopen(): fopen
memmem(): System's
times(2) sysconf(_SC_CLK_TCK) is 100
Using times(2) for timers, resolution 10 ms
HR timer: clock_gettime(), latency 29 ns
Total physical host memory: 15944 MiB
Available physical host memory: 12610 MiB
Terminal locale string: de_DE.utf8
Parsed terminal locale: UTF-8


Additional information about distribution and driver :
$ uname -a
Linux tux1 6.0.2-arch1-1 #1 SMP PREEMPT_DYNAMIC Sat, 15 Oct 2022 14:00:49 +0000 x86_64 GNU/Linux

$ pacman -Q | grep nvidia
nvidia 520.56.06-4
nvidia-settings 520.56.06-1
nvidia-utils 520.56.06-2
opencl-nvidia 520.56.06-2

$ pacman -Q | grep cuda
cuda 11.8.0-1


Similar to this (fixed)
https://github.com/openwall/john/issues/4667

Looks like NVIDA changed some API calls from time to time.

This fixes most errors:
```
$ git diff --stat
 run/opencl/ed25519-donna/curve25519-donna-32bit.h |  6 +++---
 run/opencl/opencl_des.h                           | 38 ++++++++++++++------------------------
 run/opencl/opencl_md4_ctx.h                       |  4 ++--
 run/opencl/opencl_md5_ctx.h                       |  4 ++--
 run/opencl/opencl_sha1_ctx.h                      |  4 ++--
 run/opencl/opencl_sha2_ctx.h                      |  8 ++++----
 run/opencl/opencl_twofish.h                       |  8 ++++----
 7 files changed, 31 insertions(+), 41 deletions(-)
```
That's 100% things like this:
```diff
diff --git a/run/opencl/opencl_des.h b/run/opencl/opencl_des.h
index acbcac5ad..698654c25 100644
--- a/run/opencl/opencl_des.h
+++ b/run/opencl/opencl_des.h
@@ -296,7 +296,7 @@ __constant uchar odd_parity_table[128] = { 1,  2,  4,  7,  8,
        227, 229, 230, 233, 234, 236, 239, 241, 242, 244, 247, 248, 251, 253,
        254 };
 
-inline void des_key_set_parity(uchar key[DES_KEY_SIZE])
+inline void des_key_set_parity(uchar *key)
 {
        int i;
 
```
However, now I'm hitting more complex issues:
```
In file included from opencl/ed25519-donna/ed25519-donna.c:7:
In file included from opencl/ed25519-donna/ed25519-donna.h:42:
opencl/ed25519-donna/ed25519-donna-impl-base.h:7:17: error: passing '__generic uint32_t *' (aka '__generic unsigned int *') to parameter of type 'uint32_t *' (aka 'unsigned int *') changes address space of pointer
        curve25519_mul(r->x, p->x, p->t);
                       ^~~~
opencl/ed25519-donna/curve25519-donna-32bit.h:154:28: note: passing argument to parameter 'out' here
curve25519_mul(bignum25519 out, const bignum25519 a, const bignum25519 b) {
                           ^
```

Also, I'm not glad to push the current fixes - they make the code slightly worse. This is a driver problem.

--- a/src/opencl_common.c
+++ b/src/opencl_common.c
@@ -1148,8 +1148,16 @@ static char *get_build_opts(int sequential_id, const char *opts)
 			global_opts = OPENCLBUILDOPTIONS;
 
 	snprintf(build_opts, LINE_BUFFER_SIZE,
-	         "-I opencl %s %s%s%s%s%d %s%d %s -D_OPENCL_COMPILER %s",
+	         "-I opencl %s %s%s%s%s%s%d %s%d %s -D_OPENCL_COMPILER %s",
 	        global_opts,
+	        get_device_version(sequential_id) >= 200 ? "-cl-std=CL1.2 " : "",
+#ifdef __APPLE__
+	        "-D__OS_X__ ",
+#else
+	        (options.verbosity >= VERB_MAX &&
+	         gpu_nvidia(device_info[sequential_id])) ?
+	         "-cl-nv-verbose " : "",
+#endif
 	        get_platform_vendor_id(get_platform_id(sequential_id)) ==
 	         PLATFORM_MESA ? "-D__MESA__ " :
 	        get_platform_vendor_id(get_platform_id(sequential_id)) ==
@@ -1158,13 +1166,6 @@ static char *get_build_opts(int sequential_id, const char *opts)
 	         PLATFORM_BEIGNET ?
 	         "-D__BEIGNET__ " :
 	        get_device_capability(sequential_id),
-#ifdef __APPLE__
-	        "-D__OS_X__ ",
-#else
-	        (options.verbosity >= VERB_MAX &&
-	         gpu_nvidia(device_info[sequential_id])) ?
-	         "-cl-nv-verbose " : "",
-#endif
 	        get_device_type(sequential_id) == CL_DEVICE_TYPE_CPU ? "-D__CPU__ "
 	        : get_device_type(sequential_id) == CL_DEVICE_TYPE_GPU ? "-D__GPU__ " : "",
 	        "-DDEVICE_INFO=", device_info[sequential_id],

